File: adapters/agent-bridge/oar_agent_bridge/router.py
from __future__ import annotations

import json
import logging
import time
from dataclasses import dataclass
from typing import Any

from .config import LoadedConfig
from .mentions import extract_mentions
from .models import (
    AgentBridgeCheckin,
    AgentRegistration,
    BRIDGE_CHECKED_IN_EVENT,
    MESSAGE_POSTED_EVENT,
    WAKE_ARTIFACT_KIND,
    WAKE_REQUEST_EVENT,
    WakePacket,
    registration_document_id,
    wakeup_artifact_id,
    wakeup_request_key,
)
from .oar_client import OARClient, OARClientError
from .state_store import JSONStateStore
from .util import compact_text, utc_now_iso, verify_bridge_checkin_signature

LOGGER = logging.getLogger(__name__)


@dataclass(slots=True)
class PrincipalCache:
    loaded_at: float
    principals_by_handle: dict[str, dict[str, Any]]


class WakeRouter:
    def __init__(self, config: LoadedConfig, client: OARClient, state: JSONStateStore) -> None:
        if config.router is None:
            raise ValueError("router config requires a [router] section")
        self.config = config
        self.client = client
        self.state = state
        self._principal_cache = PrincipalCache(loaded_at=0.0, principals_by_handle={})

    def run_forever(self) -> None:
        router_cfg = self.config.router
        assert router_cfg is not None
        while True:
            try:
                last_event_id = self.state.last_event_id
                self._update_router_state(
                    router_last_stream_connected_at=utc_now_iso(),
                    router_stream_resume_from_event_id=last_event_id or "",
                )
                for stream_message in self.client.stream_events(types=[MESSAGE_POSTED_EVENT], last_event_id=last_event_id):
                    event = self._decode_stream_event(stream_message)
                    if event is None:
                        continue
                    event_id = str(event.get("id", "")).strip()
                    text = self.extract_message_text(event)
                    handles = extract_mentions(text)
                    self._record_message_seen(event_id=event_id, text=text, handles=handles)
                    routed_handles = self.handle_message_posted(event)
                    self._record_message_routed(event_id=event_id, handles=routed_handles)
                    if event_id:
                        self.state.last_event_id = event_id
            except Exception as exc:
                self._update_router_state(
                    router_last_stream_error_at=utc_now_iso(),
                    router_last_stream_error=f"{type(exc).__name__}: {exc}",
                )
                LOGGER.exception("Router stream loop failed; reconnecting")
                time.sleep(router_cfg.reconnect_delay_seconds)

    def _decode_stream_event(self, stream_message: dict[str, Any]) -> dict[str, Any] | None:
        data = stream_message.get("data")
        if not isinstance(data, str) or not data.strip():
            return None
        payload = json.loads(data)
        if isinstance(payload, dict) and "event" in payload and isinstance(payload["event"], dict):
            return payload["event"]
        if isinstance(payload, dict):
            return payload
        return None

    def _update_router_state(self, **updates: Any) -> None:
        self.state.update({key: value for key, value in updates.items() if value is not None})

    def _record_message_seen(self, *, event_id: str, text: str, handles: list[str]) -> None:
        updates: dict[str, Any] = {
            "router_last_message_seen_at": utc_now_iso(),
            "router_last_message_seen_event_id": event_id,
        }
        if handles:
            updates.update(
                {
                    "router_last_tagged_message_event_id": event_id,
                    "router_last_tagged_message_seen_at": utc_now_iso(),
                    "router_last_tagged_handles": handles,
                    "router_last_tagged_message_preview": compact_text(text, 140),
                }
            )
        self._update_router_state(**updates)

    def _record_message_routed(self, *, event_id: str, handles: list[str]) -> None:
        if not event_id or not handles:
            return
        self._update_router_state(
            router_last_routed_event_id=event_id,
            router_last_routed_at=utc_now_iso(),
            router_last_routed_handles=handles,
        )

    def _load_principals(self, force: bool = False) -> None:
        ttl = self.config.router.principal_cache_ttl_seconds if self.config.router else 60
        if not force and (time.time() - self._principal_cache.loaded_at) < ttl:
            return
        principals = self.client.list_principals(limit=200)
        mapping: dict[str, dict[str, Any]] = {}
        for principal in principals:
            if principal.get("revoked"):
                continue
            if str(principal.get("principal_kind", "")).strip() != "agent":
                continue
            username = str(principal.get("username", "")).strip()
            if username:
                mapping[username] = principal
        self._principal_cache = PrincipalCache(loaded_at=time.time(), principals_by_handle=mapping)

    def handle_message_posted(self, event: dict[str, Any]) -> list[str]:
        text = self.extract_message_text(event)
        handles = extract_mentions(text)
        if not handles:
            return []
        thread_id = str(event.get("thread_id", "")).strip()
        event_id = str(event.get("id", "")).strip()
        if not thread_id or not event_id:
            LOGGER.debug("Ignoring message_posted without thread_id or id: %s", event)
            return []
        self._load_principals()
        routed_handles: list[str] = []
        for handle in handles:
            try:
                if self._route_mention(handle=handle, event=event, text=text):
                    routed_handles.append(handle)
            except Exception:
                LOGGER.exception("Failed routing mention @%s from event %s", handle, event_id)
                self._emit_exception(thread_id, event_id, handle, "mention_routing_failed", f"Failed routing @%s" % handle)
        return routed_handles

    def _route_mention(self, *, handle: str, event: dict[str, Any], text: str) -> bool:
        thread_id = str(event["thread_id"])
        event_id = str(event["id"])
        principal = self._principal_cache.principals_by_handle.get(handle)
        if principal is None:
            self._emit_exception(thread_id, event_id, handle, "unknown_agent_handle", f"Unknown tagged agent @{handle}")
            return False
        registration = self._load_registration(handle)
        if registration is None:
            self._emit_exception(thread_id, event_id, handle, "missing_agent_registration", f"Tagged agent @{handle} has no registration document")
            return False
        if registration.actor_id != str(principal.get("actor_id", "")).strip():
            self._emit_exception(thread_id, event_id, handle, "registration_actor_mismatch", f"Tagged agent @{handle} registration actor does not match principal")
            return False
        if not registration.supports_workspace(self.config.oar.workspace_id):
            self._emit_exception(thread_id, event_id, handle, "agent_not_bound_to_workspace", f"Tagged agent @{handle} is not enabled for workspace {self.config.oar.workspace_id}")
            return False
        if registration.status != "active":
            self._emit_exception(
                thread_id,
                event_id,
                handle,
                "agent_bridge_not_ready",
                f"Tagged agent @{handle} is registered but not wakeable until its bridge checks in",
            )
            return False
        if not registration.bridge_checkin_event_id:
            self._emit_exception(
                thread_id,
                event_id,
                handle,
                "agent_bridge_not_checked_in",
                f"Tagged agent @{handle} has no bridge check-in event yet",
            )
            return False
        checkin = self._load_bridge_checkin(registration.bridge_checkin_event_id)
        if checkin is None:
            self._emit_exception(
                thread_id,
                event_id,
                handle,
                "agent_bridge_not_checked_in",
                f"Tagged agent @{handle} has no valid bridge check-in event yet",
            )
            return False
        if checkin.handle and checkin.handle != handle:
            self._emit_exception(
                thread_id,
                event_id,
                handle,
                "agent_bridge_handle_mismatch",
                f"Tagged agent @{handle} bridge check-in handle does not match registration",
            )
            return False
        if not registration.bridge_signing_public_key_spki_b64:
            self._emit_exception(
                thread_id,
                event_id,
                handle,
                "agent_bridge_proof_missing",
                f"Tagged agent @{handle} registration is missing its bridge proof key",
            )
            return False
        if not verify_bridge_checkin_signature(
            registration.bridge_signing_public_key_spki_b64,
            checkin.proof_signature_b64,
            checkin.handle,
            checkin.actor_id,
            checkin.workspace_id,
            checkin.bridge_instance_id,
            checkin.checked_in_at,
            checkin.expires_at,
        ):
            self._emit_exception(
                thread_id,
                event_id,
                handle,
                "agent_bridge_proof_invalid",
                f"Tagged agent @{handle} has an invalid bridge readiness proof",
            )
            return False
        if checkin.actor_id != registration.actor_id:
            self._emit_exception(
                thread_id,
                event_id,
                handle,
                "agent_bridge_actor_mismatch",
                f"Tagged agent @{handle} bridge check-in actor does not match registration actor",
            )
            return False
        if not checkin.is_ready_for_workspace(self.config.oar.workspace_id):
            self._emit_exception(
                thread_id,
                event_id,
                handle,
                "agent_bridge_checkin_stale",
                f"Tagged agent @{handle} has a stale bridge check-in and is not wakeable right now",
            )
            return False

        workspace = self.client.get_thread_workspace(thread_id)
        thread = workspace.get("thread") or {}
        wake_artifact_id = wakeup_artifact_id(self.config.oar.workspace_id, thread_id, event_id, registration.actor_id)
        session_key = f"oar:{self.config.oar.workspace_id}:{thread_id}:{handle}"
        packet = WakePacket(
            wakeup_id=wake_artifact_id,
            handle=handle,
            actor_id=registration.actor_id,
            workspace_id=self.config.oar.workspace_id,
            workspace_name=self.config.oar.workspace_name,
            thread_id=thread_id,
            thread_title=str(thread.get("title", thread_id)).strip() or thread_id,
            trigger_event_id=event_id,
            trigger_created_at=str(event.get("ts", "")).strip(),
            trigger_author_actor_id=str(event.get("actor_id", "")).strip(),
            trigger_text=text,
            current_summary=str(thread.get("current_summary", "")).strip(),
            session_key=session_key,
            oar_base_url=self.config.oar.base_url,
            thread_context_url=f"{self.config.oar.base_url}/threads/{thread_id}/context",
            thread_workspace_url=f"{self.config.oar.base_url}/threads/{thread_id}/workspace",
            trigger_event_url=f"{self.config.oar.base_url}/events/{event_id}",
            cli_thread_inspect=f"oar threads inspect --thread-id {thread_id} --json",
            cli_thread_workspace=f"oar threads workspace --thread-id {thread_id} --include-related-event-content --json",
        )
        artifact = {
            "id": wake_artifact_id,
            "kind": WAKE_ARTIFACT_KIND,
            "summary": f"Wake packet for @{handle}",
            "refs": [f"thread:{thread_id}", f"event:{event_id}"],
            "target_handle": handle,
            "target_actor_id": registration.actor_id,
            "workspace_id": self.config.oar.workspace_id,
            "thread_id": thread_id,
        }
        try:
            self.client.create_artifact(artifact=artifact, content=packet.to_content(), content_type="structured")
        except OARClientError as exc:
            if exc.status_code != 409:
                raise
            LOGGER.info("Wake artifact %s already exists; reusing it", wake_artifact_id)
        event_payload = {
            "wakeup_id": wake_artifact_id,
            "wake_artifact_id": wake_artifact_id,
            "target_handle": handle,
            "target_actor_id": registration.actor_id,
            "workspace_id": self.config.oar.workspace_id,
            "workspace_name": self.config.oar.workspace_name,
            "thread_id": thread_id,
            "trigger_event_id": event_id,
            "session_key": session_key,
        }
        self.client.create_event(
            event={
                "type": WAKE_REQUEST_EVENT,
                "thread_id": thread_id,
                "summary": f"Wake requested for @{handle}",
                "refs": [f"thread:{thread_id}", f"event:{event_id}", f"artifact:{wake_artifact_id}"],
                "payload": event_payload,
                "provenance": {"sources": [f"actor_statement:{event_id}"]},
            },
            request_key=wakeup_request_key(self.config.oar.workspace_id, thread_id, event_id, registration.actor_id),
        )
        LOGGER.info("Queued wakeup %s for @%s in thread %s", wake_artifact_id, handle, thread_id)
        return True

    def _load_registration(self, handle: str) -> AgentRegistration | None:
        doc_id = registration_document_id(handle)
        try:
            payload = self.client.get_document(doc_id)
        except OARClientError as exc:
            if exc.status_code == 404:
                return None
            raise
        revision = payload.get("revision") or {}
        content = revision.get("content") or {}
        if not isinstance(content, dict):
            return None
        return AgentRegistration.from_content(content)

    def _load_bridge_checkin(self, event_id: str) -> AgentBridgeCheckin | None:
        try:
            payload = self.client.get_event(event_id)
        except OARClientError as exc:
            if exc.status_code == 404:
                return None
            raise
        event = payload.get("event") if isinstance(payload, dict) else None
        content = (event or {}).get("payload") if isinstance(event, dict) else None
        if str((event or {}).get("type", "")).strip() != BRIDGE_CHECKED_IN_EVENT:
            return None
        if not isinstance(content, dict):
            return None
        return AgentBridgeCheckin.from_content(content)

    def _emit_exception(self, thread_id: str, event_id: str, handle: str, code: str, summary: str) -> None:
        request_key = f"exc-{code}-{handle}-{event_id}"
        self.client.create_event(
            event={
                "type": "exception_raised",
                "thread_id": thread_id,
                "summary": summary,
                "refs": [f"thread:{thread_id}", f"event:{event_id}"],
                "payload": {"subtype": code, "code": code, "handle": handle},
                "provenance": {"sources": [f"actor_statement:{event_id}"]},
            },
            request_key=request_key,
        )

    @staticmethod
    def extract_message_text(event: dict[str, Any]) -> str:
        payload = event.get("payload") or {}
        if isinstance(payload, dict):
            for key in ("text", "message", "body", "content"):
                value = payload.get(key)
                if isinstance(value, str) and value.strip():
                    return value
        body = event.get("body")
        if isinstance(body, str) and body.strip():
            return body
        summary = event.get("summary")
        if isinstance(summary, str):
            return summary
        return ""

File: core/internal/router/router.go
package router

import (
	"context"
	"crypto/sha256"
	"encoding/hex"
	"errors"
	"fmt"
	"log"
	"strings"
	"sync"
	"time"

	"organization-autorunner-core/internal/auth"
	"organization-autorunner-core/internal/primitives"
)

const (
	defaultPollInterval = time.Second
	eventBatchSize      = 100
)

type Config struct {
	BaseURL           string
	WorkspaceID       string
	WorkspaceName     string
	StatePath         string
	PrincipalCacheTTL time.Duration
	PollInterval      time.Duration
	ActorID           string
}

type Dependencies struct {
	ListPrincipals         func(ctx context.Context, limit int) ([]auth.AuthPrincipalSummary, error)
	ListMessagePostedAfter func(ctx context.Context, cursor primitives.EventCursor, limit int) ([]map[string]any, error)
	GetEvent               func(ctx context.Context, eventID string) (map[string]any, error)
	GetThread              func(ctx context.Context, threadID string) (map[string]any, error)
	CreateArtifact         func(ctx context.Context, actorID string, artifact map[string]any, content any, contentType string) error
	AppendEvent            func(ctx context.Context, actorID string, event map[string]any) error
	MarkThreadDirty        func(ctx context.Context, threadID string, queuedAt time.Time) error
}

type principalCache struct {
	loadedAt time.Time
	byHandle map[string]auth.AuthPrincipalSummary
}

type Service struct {
	cfg   Config
	deps  Dependencies
	state *StateStore

	cache principalCache

	mu         sync.RWMutex
	ready      bool
	lastError  string
	lastPolled string
}

func NewService(cfg Config, deps Dependencies, state *StateStore) *Service {
	if cfg.PrincipalCacheTTL <= 0 {
		cfg.PrincipalCacheTTL = time.Minute
	}
	if cfg.PollInterval <= 0 {
		cfg.PollInterval = defaultPollInterval
	}
	if strings.TrimSpace(cfg.WorkspaceName) == "" {
		cfg.WorkspaceName = "Main"
	}
	if strings.TrimSpace(cfg.ActorID) == "" {
		cfg.ActorID = "oar-core"
	}
	return &Service{
		cfg:   cfg,
		deps:  deps,
		state: state,
		cache: principalCache{byHandle: map[string]auth.AuthPrincipalSummary{}},
	}
}

func (s *Service) Name() string {
	return "router"
}

func (s *Service) Run(ctx context.Context) error {
	ticker := time.NewTicker(s.cfg.PollInterval)
	defer ticker.Stop()

	for {
		if err := s.runOnce(ctx); err != nil {
			s.recordFailure(err)
			log.Printf("router sidecar poll failed: %v", err)
		} else {
			s.recordSuccess()
		}

		select {
		case <-ctx.Done():
			return ctx.Err()
		case <-ticker.C:
		}
	}
}

func (s *Service) Ready(context.Context) error {
	s.mu.RLock()
	defer s.mu.RUnlock()
	if strings.TrimSpace(s.lastError) != "" {
		return errors.New(s.lastError)
	}
	if !s.ready {
		return errors.New("initial poll not completed")
	}
	return nil
}

func (s *Service) Snapshot(context.Context) map[string]any {
	s.mu.RLock()
	ready := s.ready
	lastError := s.lastError
	lastPolled := s.lastPolled
	s.mu.RUnlock()

	snapshot := s.state.Snapshot()
	snapshot["ready"] = ready
	if strings.TrimSpace(lastError) != "" {
		snapshot["last_error"] = lastError
	}
	if strings.TrimSpace(lastPolled) != "" {
		snapshot["last_polled_at"] = lastPolled
	}
	return snapshot
}

func (s *Service) runOnce(ctx context.Context) error {
	cursor, err := s.resolveCursor(ctx)
	if err != nil {
		return err
	}

	for {
		events, err := s.deps.ListMessagePostedAfter(ctx, cursor, eventBatchSize)
		if err != nil {
			return err
		}
		if len(events) == 0 {
			return nil
		}
		for _, event := range events {
			if err := s.handleEvent(ctx, event); err != nil {
				return err
			}
			cursor = primitives.EventCursor{
				TS: anyString(event["ts"]),
				ID: anyString(event["id"]),
			}
			if err := s.state.SetLastEventCursor(cursor.TS, cursor.ID); err != nil {
				return err
			}
		}
		if len(events) < eventBatchSize {
			return nil
		}
	}
}

func (s *Service) resolveCursor(ctx context.Context) (primitives.EventCursor, error) {
	cursor := primitives.EventCursor{
		TS: s.state.LastEventTS(),
		ID: s.state.LastEventID(),
	}
	if strings.TrimSpace(cursor.TS) != "" || strings.TrimSpace(cursor.ID) == "" {
		return cursor, nil
	}
	event, err := s.deps.GetEvent(ctx, cursor.ID)
	if err != nil {
		if errors.Is(err, primitives.ErrNotFound) {
			return primitives.EventCursor{}, nil
		}
		return primitives.EventCursor{}, err
	}
	cursor.TS = anyString(event["ts"])
	return cursor, nil
}

func (s *Service) handleEvent(ctx context.Context, event map[string]any) error {
	eventID := anyString(event["id"])
	text := extractMessageText(event)
	handles := ExtractMentions(text)
	updates := map[string]any{
		"router_last_message_seen_at":       utcNowISO(),
		"router_last_message_seen_event_id": eventID,
	}
	if len(handles) > 0 {
		updates["router_last_tagged_message_event_id"] = eventID
		updates["router_last_tagged_message_seen_at"] = utcNowISO()
		updates["router_last_tagged_handles"] = handles
		updates["router_last_tagged_message_preview"] = compactText(text, 140)
	}
	if err := s.state.Update(updates); err != nil {
		return err
	}
	if len(handles) == 0 {
		return nil
	}
	if err := s.loadPrincipals(ctx, false); err != nil {
		return err
	}
	routed := make([]string, 0, len(handles))
	for _, handle := range handles {
		ok, err := s.routeMention(ctx, handle, event, text)
		if err != nil {
			log.Printf("router sidecar failed routing @%s from event %s: %v", handle, eventID, err)
			if emitErr := s.emitException(ctx, anyString(event["thread_id"]), eventID, handle, "mention_routing_failed", fmt.Sprintf("Failed routing @%s", handle)); emitErr != nil {
				log.Printf("router sidecar failed to emit exception for @%s: %v", handle, emitErr)
			}
			continue
		}
		if ok {
			routed = append(routed, handle)
		}
	}
	if len(routed) == 0 {
		return nil
	}
	return s.state.Update(map[string]any{
		"router_last_routed_event_id": eventID,
		"router_last_routed_at":       utcNowISO(),
		"router_last_routed_handles":  routed,
	})
}

func (s *Service) loadPrincipals(ctx context.Context, force bool) error {
	if !force && time.Since(s.cache.loadedAt) < s.cfg.PrincipalCacheTTL && len(s.cache.byHandle) > 0 {
		return nil
	}
	principals, err := s.deps.ListPrincipals(ctx, 200)
	if err != nil {
		return err
	}
	byHandle := make(map[string]auth.AuthPrincipalSummary, len(principals))
	for _, principal := range principals {
		if principal.Revoked || strings.TrimSpace(principal.PrincipalKind) != "agent" {
			continue
		}
		username := strings.ToLower(strings.TrimSpace(principal.Username))
		if username == "" {
			continue
		}
		byHandle[username] = principal
	}
	s.cache.loadedAt = time.Now().UTC()
	s.cache.byHandle = byHandle
	return nil
}

func (s *Service) refreshPrincipal(ctx context.Context, handle string) (auth.AuthPrincipalSummary, bool, error) {
	if err := s.loadPrincipals(ctx, true); err != nil {
		return auth.AuthPrincipalSummary{}, false, err
	}
	principal, ok := s.cache.byHandle[handle]
	return principal, ok, nil
}

func routingRegistrationFailure(principal auth.AuthPrincipalSummary, workspaceID string) (*auth.AgentRegistration, string, string) {
	registration := principal.Registration
	if registration == nil {
		return nil, "missing_agent_registration", fmt.Sprintf("Tagged agent @%s has no registration", principal.Username)
	}
	if strings.TrimSpace(registration.ActorID) != strings.TrimSpace(principal.ActorID) {
		return nil, "registration_actor_mismatch", fmt.Sprintf("Tagged agent @%s registration actor does not match principal", principal.Username)
	}
	if !registration.SupportsWorkspace(workspaceID) {
		return nil, "agent_not_bound_to_workspace", fmt.Sprintf("Tagged agent @%s is not enabled for workspace %s", principal.Username, workspaceID)
	}
	if strings.EqualFold(strings.TrimSpace(registration.Status), "disabled") {
		return nil, "agent_notifications_disabled", fmt.Sprintf("Tagged agent @%s is disabled for notifications", principal.Username)
	}
	return registration, "", ""
}

func (s *Service) routeMention(ctx context.Context, handle string, event map[string]any, text string) (bool, error) {
	threadID := anyString(event["thread_id"])
	eventID := anyString(event["id"])
	if threadID == "" || eventID == "" {
		return false, nil
	}

	principal, ok := s.cache.byHandle[handle]
	if !ok {
		refreshed, refreshedOK, err := s.refreshPrincipal(ctx, handle)
		if err != nil {
			return false, err
		}
		if !refreshedOK {
			return false, s.emitException(ctx, threadID, eventID, handle, "unknown_agent_handle", fmt.Sprintf("Unknown tagged agent @%s", handle))
		}
		principal = refreshed
	}

	registration, code, message := routingRegistrationFailure(principal, s.cfg.WorkspaceID)
	if code != "" {
		refreshed, refreshedOK, err := s.refreshPrincipal(ctx, handle)
		if err != nil {
			return false, err
		}
		if !refreshedOK {
			return false, s.emitException(ctx, threadID, eventID, handle, "unknown_agent_handle", fmt.Sprintf("Unknown tagged agent @%s", handle))
		}
		principal = refreshed
		registration, code, message = routingRegistrationFailure(principal, s.cfg.WorkspaceID)
		if code != "" {
			return false, s.emitException(ctx, threadID, eventID, handle, code, message)
		}
	}
	if strings.TrimSpace(anyString(event["actor_id"])) == registration.ActorID {
		return false, nil
	}

	thread, err := s.deps.GetThread(ctx, threadID)
	if err != nil {
		return false, err
	}
	subjectRef, resolvedSubject := ResolvedSubjectFromThread(thread, threadID)
	baseURL := strings.TrimRight(s.cfg.BaseURL, "/")
	topicWorkspaceURL := ""
	cliTopicWorkspace := ""
	if sr := strings.TrimSpace(subjectRef); strings.HasPrefix(sr, "topic:") {
		topicID := strings.TrimSpace(strings.TrimPrefix(sr, "topic:"))
		if topicID != "" {
			topicWorkspaceURL = fmt.Sprintf("%s/topics/%s/workspace", baseURL, topicID)
			cliTopicWorkspace = fmt.Sprintf("oar topics workspace --topic-id %s --json", topicID)
		}
	}
	wakeupID := WakeupArtifactID(s.cfg.WorkspaceID, threadID, eventID, registration.ActorID)
	sessionKey := fmt.Sprintf("oar:%s:%s:%s", s.cfg.WorkspaceID, threadID, handle)
	packet := WakePacket{
		WakeupID:             wakeupID,
		Handle:               handle,
		ActorID:              registration.ActorID,
		WorkspaceID:          s.cfg.WorkspaceID,
		WorkspaceName:        s.cfg.WorkspaceName,
		ThreadID:             threadID,
		ThreadTitle:          firstNonEmpty(anyString(thread["title"]), threadID),
		SubjectRef:           subjectRef,
		ResolvedSubject:      resolvedSubject,
		TriggerEventID:       eventID,
		TriggerCreatedAt:     anyString(event["ts"]),
		TriggerAuthorActorID: anyString(event["actor_id"]),
		TriggerText:          text,
		CurrentSummary:       anyString(thread["current_summary"]),
		SessionKey:           sessionKey,
		OARBaseURL:           baseURL,
		ThreadContextURL:     fmt.Sprintf("%s/threads/%s/context", baseURL, threadID),
		ThreadWorkspaceURL:   fmt.Sprintf("%s/threads/%s/workspace", baseURL, threadID),
		TopicWorkspaceURL:    topicWorkspaceURL,
		TriggerEventURL:      fmt.Sprintf("%s/events/%s", baseURL, eventID),
		CLIThreadInspect:     fmt.Sprintf("oar threads inspect --thread-id %s --json", threadID),
		CLIThreadWorkspace:   fmt.Sprintf("oar threads workspace --thread-id %s --include-related-event-content --json", threadID),
		CLITopicWorkspace:    cliTopicWorkspace,
		Version:              WakePacketVersion,
	}

	wakeRefs := append(WakeArtifactRefs(threadID, eventID, subjectRef), fmt.Sprintf("artifact:%s", wakeupID))
	artifact := map[string]any{
		"id":              wakeupID,
		"kind":            WakeArtifactKind,
		"summary":         fmt.Sprintf("Wake packet for @%s", handle),
		"refs":            WakeArtifactRefs(threadID, eventID, subjectRef),
		"target_handle":   handle,
		"target_actor_id": registration.ActorID,
		"workspace_id":    s.cfg.WorkspaceID,
		"thread_id":       threadID,
	}
	if err := s.deps.CreateArtifact(ctx, s.cfg.ActorID, artifact, packet.ToContent(), "structured"); err != nil && !errors.Is(err, primitives.ErrConflict) {
		return false, err
	}

	requestKey := WakeupRequestKey(s.cfg.WorkspaceID, threadID, eventID, registration.ActorID)
	eventBody := map[string]any{
		"type":      WakeRequestEvent,
		"thread_id": threadID,
		"summary":   fmt.Sprintf("Wake requested for @%s", handle),
		"refs":      wakeRefs,
		"payload": BuildWakeRequestPayload(
			wakeupID, handle, registration.ActorID,
			s.cfg.WorkspaceID, s.cfg.WorkspaceName, threadID,
			eventID, anyString(event["ts"]), text, sessionKey,
			subjectRef, resolvedSubject,
		),
		"provenance": map[string]any{
			"sources": []string{fmt.Sprintf("actor_statement:%s", eventID)},
		},
	}
	if err := s.appendThreadEvent(ctx, requestKey, eventBody); err != nil {
		return false, err
	}
	return true, nil
}

func (s *Service) emitException(ctx context.Context, threadID string, eventID string, handle string, code string, summary string) error {
	requestKey := fmt.Sprintf("exc-%s-%s-%s", code, handle, eventID)
	return s.appendThreadEvent(ctx, requestKey, map[string]any{
		"type":      "exception_raised",
		"thread_id": threadID,
		"summary":   summary,
		"refs":      []string{fmt.Sprintf("thread:%s", threadID), fmt.Sprintf("event:%s", eventID)},
		"payload": map[string]any{
			"subtype": code,
			"code":    code,
			"handle":  handle,
		},
		"provenance": map[string]any{
			"sources": []string{fmt.Sprintf("actor_statement:%s", eventID)},
		},
	})
}

func (s *Service) appendThreadEvent(ctx context.Context, requestKey string, event map[string]any) error {
	if strings.TrimSpace(requestKey) != "" && strings.TrimSpace(anyString(event["id"])) == "" {
		event["id"] = deriveRequestScopedID("events.create", s.cfg.ActorID, requestKey, "event")
	}
	err := s.deps.AppendEvent(ctx, s.cfg.ActorID, event)
	if err != nil {
		if errors.Is(err, primitives.ErrConflict) {
			return nil
		}
		return err
	}
	threadID := anyString(event["thread_id"])
	if strings.TrimSpace(threadID) != "" && s.deps.MarkThreadDirty != nil {
		if err := s.deps.MarkThreadDirty(ctx, threadID, nowUTC()); err != nil {
			return err
		}
	}
	return nil
}

func (s *Service) recordSuccess() {
	s.mu.Lock()
	defer s.mu.Unlock()
	s.ready = true
	s.lastError = ""
	s.lastPolled = utcNowISO()
}

func (s *Service) recordFailure(err error) {
	s.mu.Lock()
	defer s.mu.Unlock()
	s.lastError = err.Error()
	s.lastPolled = utcNowISO()
}

func deriveRequestScopedID(scope string, actorID string, requestKey string, label string) string {
	sum := sha256.Sum256([]byte(scope + "\n" + actorID + "\n" + requestKey + "\n" + label))
	return label + "-" + hex.EncodeToString(sum[:])[:20]
}
